exchanges/okx/okx_test.go

```diff
@@ -0,0 +1,3180 @@
+package okx
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"log"
+	"os"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"github.com/thrasher-corp/gocryptotrader/config"
+	"github.com/thrasher-corp/gocryptotrader/core"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	exchange "github.com/thrasher-corp/gocryptotrader/exchanges"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/kline"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/order"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/orderbook"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/request"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/sharedtestvalues"
+	"github.com/thrasher-corp/gocryptotrader/portfolio/withdraw"
+)
+
+// Please supply your own keys here to do authenticated endpoint testing
+const (
+	apiKey                  = ""
+	apiSecret               = ""
+	passphrase              = ""
+	canManipulateRealOrders = false
+)
+
+var ok Okx
+
+func TestMain(m *testing.M) {
+	cfg := config.GetConfig()
+	err := cfg.LoadConfig("../../testdata/configtest.json", true)
+	if err != nil {
+		log.Fatal(err)
+	}
+	exchCfg, err := cfg.GetExchangeConfig("Okx")
+	if err != nil {
+		log.Fatal(err)
+	}
+	exchCfg.API.Credentials.Key = apiKey
+	exchCfg.API.Credentials.Secret = apiSecret
+	exchCfg.API.Credentials.ClientID = passphrase
+	ok.WsResponseMultiplexer = wsRequestDataChannelsMultiplexer{
+		WsResponseChannelsMap: make(map[string]*wsRequestInfo),
+		Register:              make(chan *wsRequestInfo),
+		Unregister:            make(chan string),
+		Message:               make(chan *wsIncomingData),
+	}
+	ok.SetDefaults()
+	if apiKey != "" && apiSecret != "" {
+		exchCfg.API.AuthenticatedSupport = true
+		exchCfg.API.AuthenticatedWebsocketSupport = true
+	}
+	err = ok.Setup(exchCfg)
+	if err != nil {
+		log.Fatal(err)
+	}
+	ok.Websocket = sharedtestvalues.NewTestWebsocket()
+	request.MaxRequestJobs = 200
+	err = ok.Setup(exchCfg)
```

Comment:
Can rid this second call to `ok.Setup(exchCfg)` and move the ok.Websocket = sharedtestvalues.NewTestWebsocket() to above the first call to Setup()